### ISSUE TITLE
precompilation safety

### ISSUE THREAD

**Issue by stevengj** [JuliaGL/GLFW.jl#53]
If you ever precompile this package, e.g. as a side-effect of precompiling GLVisualize (JuliaGL/GLVisualize.jl#15), it should be revised to separate runtime and compile-time more cleanly.

In particular, it currently detects the `glfwGetVersion` at runtime and initializes appropriately.  If you precompile this module, however, it will "freeze in" the `glfwGetVersion` that was detected at compile time and probably crash if the version loaded at runtime is different. 

Since the version can no longer change at runtime anyway once it is precompiled (which is probably desirable), my suggestion is:

* Detect the glfw version and library location in `deps/build.jl`, i.e. at `Pkg.build` time, and write it to a file `deps/deps.jl` that is included by `GLFW.jl`
* At runtime (in an `__init__` function), call `glfwGetVersion` and throw an exception if the major version does not match the one in `deps.jl` (i.e. if it doesn't match the version the package was built with) — tell the caller to re-run `Pkg.build("GLFW")`.

**Comment by jayschwa** [comment]
Just so I'm clear, you're saying that it's working okay with precompilation as long as the dynamic library isn't changed? Seemed to work when I tried it.

I suppose we can still add a check with friendlier error message at runtime, though it's not clear to me when someone would actually run into this issue. I think if we make use of [`include_dependency`](https://julia.readthedocs.org/en/latest/stdlib/base/#Base.include_dependency), the problem can be avoided altogether - anytime the dynamic lib changes, it'll get recompiled. Do you think that would be satisfactory?

**Comment by SimonDanisch** [comment]
do you mean `include_dependency("path\\to\\glfw\lib")` ?
Sounds smart to me...

**Comment by stevengj** [comment]
That should work, but it will cause some unnecessary recompilation (for library updates that do not change the major version).

**Comment by jayschwa** [comment]
Library updates are fairly infrequent (generally only one or two a year), so I don't think triggering recompilation on them will cause too much churn. Recompilation is also good for minor version updates since it will unlock new features (assuming the package has been updated to support them).

I was experimenting with `include_dependency` tonight. It seems to work well, but I was hoping it would simply work with the output of `Libdl.find_library`. They don't play nice since the former expects a file extension and the latter doesn't provide one. I started a thread in julia-dev asking about the best way to do this: https://groups.google.com/d/topic/julia-dev/Xk3occIColg/discussion
